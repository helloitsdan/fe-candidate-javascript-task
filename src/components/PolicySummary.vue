<template>
  <div>
    <h3 class="is-marginless">
      {{ insured_entities | summariseInsuredEntities }}
    </h3>

    <p class="active" v-if="status === 'ACTIVE'">
      <strong>Active</strong>
    </p>

    <p class="cancelled" v-if="status === 'CANCELLED'">
      Cancelled
    </p>

    <p class="lapsed" v-if="status === 'LAPSED'">
      Lapsed
    </p>

    <p>Started on {{ start_date | moment("dddd, MMMM Do YYYY") }}</p>
  </div>
</template>

<script lang="ts">
import Vue from "vue";

export default Vue.extend({
  name: "PolicySummary",
  props: {
    uuid: String,
    status: String,
    insured_entities: Array,
    start_date: String
  }
});
</script>
