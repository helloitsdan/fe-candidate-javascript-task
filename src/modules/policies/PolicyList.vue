<template>
  <div>
    <h2 class="title-margin">Policies</h2>

    <div v-if="loading" class="loading">
      Loading...
    </div>

    <div v-if="error" class="error">
      Sorry! We couldn't load this customer's policies right now.
    </div>

    <div v-if="policies" class="columns is-multiline">
      <div
        v-for="policy in policies"
        v-bind:key="policy.uuid"
        class="column is-one-third"
      >
        <div class="card background--white">
          <policy-summary v-bind="policy" />

          <p class="is-marginless">
            <router-link
              :to="{
                name: 'PolicyDetails',
                params: { policyId: policy.uuid }
              }"
            >
              View details
            </router-link>
          </p>
        </div>
      </div>
    </div>
  </div>
</template>

<script lang="ts">
import Vue from "vue";

import getPolicies from "@/services/policies";
import PolicySummary from "@/components/PolicySummary.vue";
import compareDateStrings from "@/utils/compareDateStrings";

/* This should be a computed property, but ts3.4 has an open issue where they break type inferrence */
const sortPolicyList = (policies: [{ start_date: string }]) => {
  if (!policies) {
    return [];
  }

  return policies.sort((policyA, policyB) =>
    compareDateStrings(policyA.start_date, policyB.start_date)
  );
};

export default Vue.extend({
  name: "PolicyList",
  props: ["customerId"],
  components: {
    PolicySummary
  },
  data() {
    return {
      loading: false,
      error: null,
      policies: []
    };
  },
  created() {
    this.fetchPolicyList();
  },
  methods: {
    async fetchPolicyList() {
      const { customerId } = this.$props;

      if (this.loading) {
        return;
      }

      this.loading = true;

      try {
        this.policies = sortPolicyList(await getPolicies(customerId)) as [];
      } catch (e) {
        this.error = e;
      }

      this.loading = false;
    }
  }
});
</script>
