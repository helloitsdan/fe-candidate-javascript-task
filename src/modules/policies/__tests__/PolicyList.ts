import Vue from "vue";
import { shallowMount } from "@vue/test-utils";

import PolicyList from "../PolicyList.vue";
import PolicySummary from "@/components/PolicySummary.vue";
import policyRequest from "@/services/policies";

jest.mock("@/services/policies");

describe("PolicyList", () => {
  const policyRequestMock = policyRequest as jest.Mock;

  const policiesList = [
    {
      uuid: "1234",
      start_date: "2016-07-13T02:57:55.235Z",
      status: "ACTIVE",
      insured_entities: [
        { species: "CAT" },
        { species: "CAT" },
        { species: "DOG" }
      ]
    },
    {
      uuid: "5678",
      start_date: "2016-07-19T08:57:55.235Z",
      status: "INACTIVE",
      insured_entities: [{ species: "DOG" }]
    }
  ];

  it("renders", async () => {
    policyRequestMock.mockReturnValue(Promise.resolve(policiesList));
    const wrapper = shallowMount(PolicyList, {
      stubs: ["router-link"]
    });

    await Vue.nextTick();

    expect(wrapper.element).toMatchSnapshot();
  });

  it("renders a PolicyCard for every customer", async () => {
    policyRequestMock.mockReturnValue(Promise.resolve(policiesList));
    const wrapper = shallowMount(PolicyList, {
      stubs: ["router-link"]
    });

    await Vue.nextTick();

    expect(wrapper.findAll(PolicySummary).length).toEqual(policiesList.length);
  });

  it("renders a loading message when the request is still pending", () => {
    const wrapper = shallowMount(PolicyList, {
      stubs: ["router-link"]
    });

    expect(wrapper.find(".loading").exists()).toBe(true);
  });

  it("renders an error message when the request fails", async () => {
    policyRequestMock.mockReturnValue(Promise.reject(":("));
    const wrapper = shallowMount(PolicyList, {
      stubs: ["router-link"]
    });

    await Vue.nextTick();

    expect(wrapper.find(".error").exists()).toBe(true);
  });
});
