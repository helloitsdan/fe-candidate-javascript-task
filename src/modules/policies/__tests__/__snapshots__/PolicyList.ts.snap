// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PolicyList renders 1`] = `
<div>
  <h2
    class="title-margin"
  >
    Policies
  </h2>
   
  <!---->
   
  <!---->
   
  <div
    class="columns is-multiline"
  >
    <div
      class="column is-one-third"
    >
      <div
        class="card background--white"
      >
        <policy-summary-stub
          insured_entities="[object Object]"
          start_date="2016-07-19T08:57:55.235Z"
          status="INACTIVE"
          uuid="5678"
        />
         
        <p
          class="is-marginless"
        >
          <router-link-stub
            to="[object Object]"
          >
            
            View details
          
          </router-link-stub>
        </p>
      </div>
    </div>
    <div
      class="column is-one-third"
    >
      <div
        class="card background--white"
      >
        <policy-summary-stub
          insured_entities="[object Object],[object Object],[object Object]"
          start_date="2016-07-13T02:57:55.235Z"
          status="ACTIVE"
          uuid="1234"
        />
         
        <p
          class="is-marginless"
        >
          <router-link-stub
            to="[object Object]"
          >
            
            View details
          
          </router-link-stub>
        </p>
      </div>
    </div>
  </div>
</div>
`;
